File: qbopt/backend/narrow.py
"""Load narrowing: a wide load read only through its word halves is those words loaded.

LLVM's DAGCombiner does this for `trunc (load)` and `srl (load)`. It is
selection, not optimization: the MIR already says only the halves are read,
and a word is this target's natural load. Extracting the high word of a dword
register otherwise costs a push and two pops.
"""

from dataclasses import replace

from qbopt.model import mir

_WORD = 2
_HALVES = {0: 0, 16: _WORD}  # extract bit offset -> byte offset


def narrowed(body: mir.MirBody) -> mir.MirBody:
    """`body` with every narrowable dword load split into its read words."""
    readers: dict[mir.Value, list[mir.Op]] = {}
    for block in body.blocks:
        for op in block.ops:
            for value in op.uses:
                readers.setdefault(value, []).append(op)
    phis = {value for block in body.blocks for phi in block.phis for value in phi.incoming.values()}
    halves: dict[int, list[tuple[int, mir.Value]]] = {}
    gone: set[int] = set()
    for block in body.blocks:
        for op in block.ops:
            value = _loaded(op)
            if value is None or value in phis:
                continue
            reading = readers.get(value, [])
            offsets = [_half(one, value) for one in reading]
            if not reading or None in offsets:
                continue
            halves[id(op)] = [(offset, one.results[0].value) for offset, one in zip(offsets, reading)]
            gone.update(id(one) for one in reading)
    if not halves:
        return body
    return replace(
        body,
        blocks=tuple(
            replace(block, ops=tuple(one for op in block.ops for one in _rewritten(op, halves, gone)))
            for block in body.blocks
        ),
    )


def _loaded(op: mir.Op) -> "mir.Value | None":
    """The dword a plain load of one exactly addressed cell defines."""
    match op.kind, op.args, op.results:
        case mir.Kind.LOAD, (mir.Cell(ref=ref),), (mir.Held(value=value, width=4),):
            # A source-backed load's bytes belong to the source-map emitter.
            if (
                ref.width == 4
                and ref.addr is not None
                and not ref.volatile
                and op.loads == (ref,)
                and not op.source_backed
            ):
                return value
    return None


def _half(op: mir.Op, value: mir.Value) -> "int | None":
    """The byte offset of the word this operation extracts from `value`."""
    match op.kind, op.args, op.results:
        case mir.Kind.EXTRACT, (mir.Held(value=source), mir.Const(n=bit)), (mir.Held(width=2),):
            if source == value and bit in _HALVES and op.uses == (value,):
                return _HALVES[bit]
    return None


def _rewritten(op: mir.Op, halves: dict, gone: set[int]) -> tuple[mir.Op, ...]:
    if id(op) in gone:
        return ()
    if id(op) not in halves:
        return (op,)
    (ref,) = op.loads
    made: dict[int, mir.Value] = {}
    out = []
    for offset, result in sorted(halves[id(op)], key=lambda one: one[0]):
        if offset in made:
            out.append(
                replace(
                    op,
                    kind=mir.Kind.COPY,
                    name="",
                    defines=(result,),
                    uses=(made[offset],),
                    args=(mir.Held(made[offset], _WORD),),
                    results=(mir.Held(result, _WORD),),
                    loads=(),
                )
            )
            continue
        made[offset] = result
        word = replace(ref, addr=ref.addr.plus(offset), width=_WORD)
        out.append(
            replace(
                op,
                defines=(result,),
                args=(mir.Cell(word),),
                results=(mir.Held(result, _WORD),),
                loads=(word,),
            )
        )
    return tuple(out)

File: src/backend/narrow.rs
//! Port of `qbopt/backend/narrow.py`.
//!
//! Load narrowing: a wide load read only through its word halves is those words loaded.
//!
//! LLVM's DAGCombiner does this for `trunc (load)` and `srl (load)`. It is
//! selection, not optimization: the MIR already says only the halves are read,
//! and a word is this target's natural load. Extracting the high word of a dword
//! register otherwise costs a push and two pops.

use std::collections::BTreeSet;
use crate::support::hash::HashMap;

use crate::support::hash::IndexMap;

use crate::model::mir::{self, Arg, Kind, MirBlock, MirBody, Op, Value};

const _WORD: u32 = 2;

// extract bit offset -> byte offset
fn _halves(bit: i64) -> Option<i64> {
    match bit {
        0 => Some(0),
        16 => Some(i64::from(_WORD)),
        _ => None,
    }
}

// Python's `id(op)`: an operation's position in the body.
type Id = (usize, usize);

/// `body` with every narrowable dword load split into its read words.
pub fn narrowed(body: &MirBody) -> MirBody {
    let mut readers: HashMap<Value, Vec<Id>> = HashMap::default();
    for (b, block) in body.blocks.iter().enumerate() {
        for (o, op) in block.ops.iter().enumerate() {
            for value in &op.uses {
                readers.entry(*value).or_default().push((b, o));
            }
        }
    }
    let phis: BTreeSet<Value> =
        body.blocks.iter().flat_map(|block| &block.phis).flat_map(|phi| phi.incoming.values().copied()).collect();
    let op_at = |(b, o): Id| &body.blocks[b].ops[o];
    let mut halves: IndexMap<Id, Vec<(i64, Value)>> = IndexMap::default();
    let mut gone: BTreeSet<Id> = BTreeSet::new();
    for (b, block) in body.blocks.iter().enumerate() {
        for (o, op) in block.ops.iter().enumerate() {
            let Some(value) = _loaded(op) else {
                continue;
            };
            if phis.contains(&value) {
                continue;
            }
            let reading = readers.get(&value).cloned().unwrap_or_default();
            let offsets: Vec<Option<i64>> = reading.iter().map(|one| _half(op_at(*one), value)).collect();
            if reading.is_empty() || offsets.contains(&None) {
                continue;
            }
            halves.insert(
                (b, o),
                offsets
                    .iter()
                    .zip(&reading)
                    .map(|(offset, one)| (offset.expect("checked"), _result(op_at(*one))))
                    .collect(),
            );
            gone.extend(reading);
        }
    }
    if halves.is_empty() {
        return body.clone();
    }
    MirBody {
        blocks: body
            .blocks
            .iter()
            .enumerate()
            .map(|(b, block)| MirBlock {
                ops: block
                    .ops
                    .iter()
                    .enumerate()
                    .flat_map(|(o, op)| _rewritten(op, (b, o), &halves, &gone))
                    .collect(),
                ..block.clone()
            })
            .collect(),
        ..body.clone()
    }
}

fn _result(op: &Op) -> Value {
    match &op.results[0] {
        Arg::Held(held) => held.value,
        _ => unreachable!("_half matched a held result"),
    }
}

/// The dword a plain load of one exactly addressed cell defines.
fn _loaded(op: &Op) -> Option<Value> {
    if op.kind != Kind::Load {
        return None;
    }
    match (op.args.as_slice(), op.results.as_slice()) {
        ([Arg::Cell(mir::Cell { r#ref })], [Arg::Held(mir::Held { value, width: 4 })]) => {
            // A source-backed load's bytes belong to the source-map emitter.
            if r#ref.width == 4
                && r#ref.addr.is_some()
                && !r#ref.volatile
                && op.loads == [r#ref.clone()]
                && !op.source_backed
            {
                return Some(*value);
            }
            None
        }
        _ => None,
    }
}

/// The byte offset of the word this operation extracts from `value`.
fn _half(op: &Op, value: Value) -> Option<i64> {
    if op.kind != Kind::Extract {
        return None;
    }
    match (op.args.as_slice(), op.results.as_slice()) {
        ([Arg::Held(mir::Held { value: source, .. }), Arg::Const(mir::Const { n: bit, .. })], [Arg::Held(mir::Held { width: 2, .. })]) => {
            let offset = i64::try_from(bit).ok().and_then(_halves);
            if *source == value && offset.is_some() && op.uses == [value] {
                return offset;
            }
            None
        }
        _ => None,
    }
}

fn _rewritten(op: &Op, id: Id, halves: &IndexMap<Id, Vec<(i64, Value)>>, gone: &BTreeSet<Id>) -> Vec<Op> {
    if gone.contains(&id) {
        return Vec::new();
    }
    let Some(found) = halves.get(&id) else {
        return vec![op.clone()];
    };
    let [r#ref] = op.loads.as_slice() else {
        unreachable!("_loaded matched one load")
    };
    let mut made: HashMap<i64, Value> = HashMap::default();
    let mut out = Vec::new();
    let mut sorted = found.clone();
    sorted.sort_by_key(|one| one.0);
    for (offset, result) in sorted {
        if let Some(earlier) = made.get(&offset) {
            out.push(Op {
                kind: Kind::Copy,
                name: String::new(),
                defines: vec![result],
                uses: vec![*earlier],
                args: vec![Arg::Held(mir::Held { value: *earlier, width: _WORD })],
                results: vec![Arg::Held(mir::Held { value: result, width: _WORD })],
                loads: Vec::new(),
                ..op.clone()
            });
            continue;
        }
        made.insert(offset, result);
        let word = mir::MemRef {
            addr: r#ref.addr.map(|addr| addr.plus(offset)),
            width: _WORD,
            ..r#ref.clone()
        };
        out.push(Op {
            defines: vec![result],
            args: vec![Arg::Cell(mir::Cell { r#ref: word.clone() })],
            results: vec![Arg::Held(mir::Held { value: result, width: _WORD })],
            loads: vec![word],
            ..op.clone()
        });
    }
    out
}
